File: Data Structures & Algorithms/binary-search-tree-iterator/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.data = []

        def dfs(root):
            if not root: return None
            dfs(root.left)
            self.data.append(root.val)
            dfs(root.right)

        dfs(root)
        print(self.data)
        self.idx = 0

    def next(self) -> int:
        self.idx +=1
        return self.data[self.idx-1]

    def hasNext(self) -> bool:
        if len(self.data) > self.idx:
            return True
        else:
            return False
        


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool solve(TreeNode* root, int minval, int maxval){
        if(root == NULL) return true;
        if(root->val <= minval || root->val >= maxval) return false;
        bool l = solve(root->left, minval, min(maxval,root->val));
        bool r = solve(root->right, max(minval, root->val), maxval);
        return l && r;
    }
    bool isValidBST(TreeNode* root) {
        return solve(root, INT_MIN, INT_MAX);;
    }
};
